routers/web/auth/oauth.go

```diff
@@ -327,10 +326,29 @@ func getOAuthGroupsForUser(ctx go_context.Context, user *user_model.User) ([]str
 	return groups, nil
 }
 
+func parseBasicAuth(ctx *context.Context) (username, password string, err error) {
+	authHeader := ctx.Req.Header.Get("Authorization")
+	if authType, authData, ok := strings.Cut(authHeader, " "); ok && authType == "Basic" {
+		return base.BasicAuthDecode(authData)
+	}
+	return "", "", errors.New("invalid basic authentication")
+}
+
 // IntrospectOAuth introspects an oauth token
 func IntrospectOAuth(ctx *context.Context) {
-	if ctx.Doer == nil {
-		ctx.Resp.Header().Set("WWW-Authenticate", `Bearer realm=""`)
+	clientIDValid := false
+	if clientID, clientSecret, err := parseBasicAuth(ctx); err == nil {
+		if app, err := auth.GetOAuth2ApplicationByClientID(ctx, clientID); err == nil {
```

Comment:
Do not use `else` move error check condition here and as it uses return you can drop else and assign variable after if

Reply:
Sorry, like this?

```go
        clientIDValid := false
        if clientID, clientSecret, err := parseBasicAuth(ctx); err == nil {
                app, err := auth.GetOAuth2ApplicationByClientID(ctx, clientID)
                if err == nil {
                        clientIDValid = app.ValidateClientSecret([]byte(clientSecret))
                }
                if err != nil && !auth.IsErrOauthClientIDInvalid(err) {
                        // this is likely a database error; log it and respond without details
                        log.Error("Error retrieving client_id: %v", err)
                        ctx.Error(http.StatusInternalServerError)
                        return
                }
        }
```

The control flow is less readable this way IMO.

Reply:
No this way:

```go
        clientIDValid := false
        if clientID, clientSecret, err := parseBasicAuth(ctx); err == nil {
                app, err := auth.GetOAuth2ApplicationByClientID(ctx, clientID)
                if err != nil && !auth.IsErrOauthClientIDInvalid(err) {
                        // this is likely a database error; log it and respond without details
                        log.Error("Error retrieving client_id: %v", err)
                        ctx.Error(http.StatusInternalServerError)
                        return
                }

                clientIDValid = err == nil && app.ValidateClientSecret([]byte(clientSecret))
        }
```